vllm/model_executor/model_loader/weight_utils.py

```diff
@@ -60,6 +61,92 @@
 # system reboots, so users will not complain about annoying lock files
 temp_dir = tempfile.gettempdir()
 
+def support_quantized_model_reload(func):
+    """decorator for `load_weights` flashrl patch
+    """
+    def patched_load_weights(self, weights: Iterable[tuple[str,
+                                                   torch.Tensor]]) -> set[str]:
+        if self.quant_config is None or not hasattr(self, original_weights_rebuild_keys):
+            # after parameter loading, additionall process weights is needed 
+            setattr(model, 'process_weights_after_loading_already_called', False)
+            
+            return func(self, weights)
+        else:
+                        
+            for _, module in self.named_modules():
+                if torch.is_tensor(getattr(module, 'workspace', None)):
+                    setattr(module, f'preserved_workspace', getattr(module, 'workspace'))
+            
+            existing_params = dict(self.named_parameters())
+            
+            # preserve original data, so that after parameter loading, we can put the parameter
+            # in the original tensor
+            original_param_dict = {}
+            for name, p in existing_params.items():
+                original_param_dict[name] = p.data
+                
+            # recover the parameter to the state before first loading
+            for name, (shape, stride, dtype, nbytes) in self.original_weights_rebuild_keys.items():
+                if name in existing_params:
+                    existing_params[name].data = torch.empty(shape, dtype=dtype) 
+            
+            for k, loader_k in self.recorded_loader.items():
+                for n, loader in loader_k.items():
+                    if not hasattr(existing_params[n], k):
+                        setattr(existing_params[n], k, bond_method_to_cls(loader, existing_params[n]))
+
+            updated_params = func(self, weights)
+            
+            # manually conducting process weights after loading
+            process_weights_after_loading(self, None, None)
+            setattr(self, 'process_weights_after_loading_already_called', True)
+            
+            # if quantization is the naive `fp8`, the updated_params will only include `weight` while
+            # `weight_scale` is also updated but not included in the updated_params. 
+            if self.quant_config.get_name() == 'fp8' and (not self.quant_config.is_checkpoint_fp8_serialized):
```

Comment:
should we have a more general API for all quantization method, instead of hardcode for a specific one?

Reply:
Thanks for the comment! For clarification, the API here is designed for all quantization method in general. As to the`fp8` hardcode here, it is because `fp8` is the only (to the best of my knowledge) native quantization method supported by vLLM, i.e., the quantization will be performed by vLLM & all `weight-scale` are computed by vLLM instead of being passed by user.

If needed, we can add another utitlity function to identity whether the user-specified quantization is carried out by vLLM. Let me know which approach is preferred!